test/test.downscopedclient.ts

```diff
@@ -0,0 +1,406 @@
+
+import {StsSuccessfulResponse} from '../src/auth/stscredentials';
+import {DownscopedClient} from '../src/auth/downscopedclient';
+import {AuthClient} from '../src/auth/authclient';
+import {mockStsBetaTokenExchange} from './externalclienthelper';
+import {GoogleAuth} from '../src';
+
+nock.disableNetConnect();
+
+interface SampleResponse {
+  foo: string;
+  bar: number;
+}
+
+describe('DownscopedClient', () => {
+  let clock: sinon.SinonFakeTimers;
+
+  const auth = new GoogleAuth({
+    keyFilename: './test/fixtures/private.json',
+    scopes: 'https://www.googleapis.com/auth/cloud-platform',
+  });
+  let client: AuthClient;
+
+  const ONE_HOUR_IN_SECS = 3600;
+  const testAvailableResource =
+    '//storage.googleapis.com/projects/_/buckets/bucket-123';
+  const testAvailablePermission1 = 'inRole:roles/storage.objectViewer';
+  const testAvailablePermission2 = 'inRole:roles/storage.objectAdmin';
+  const testAvailabilityConditionExpression =
+    "resource.name.startsWith('projects/_/buckets/bucket-123/objects/prefix')";
+  const testAvailabilityConditionTitle = 'Test availability condition title.';
+  const testAvailabilityConditionDescription =
+    'Test availability condition description.';
+  const testClientAccessBoundary = {
+    accessBoundary: {
+      accessBoundaryRules: [
+        {
+          availableResource: testAvailableResource,
+          availablePermissions: [testAvailablePermission1],
+          availabilityCondition: {
+            expression: testAvailabilityConditionExpression,
+          },
+        },
+      ],
+    },
+  };
+  const stsSuccessfulResponse: StsSuccessfulResponse = {
+    access_token: 'DOWNSCOPED_CLIENT_ACCESS_TOKEN_0',
+    issued_token_type: 'urn:ietf:params:oauth:token-type:access_token',
+    token_type: 'Bearer',
+    expires_in: ONE_HOUR_IN_SECS,
+    scope: 'scope1 scope2',
+  };
+  /**
+   * Offset to take into account network delays and server clock skews.
+   */
+  const EXPIRATION_TIME_OFFSET = 5 * 60 * 1000;
+
+  beforeEach(async () => {
+    client = await auth.getClient();
+  });
+
+  afterEach(() => {
+    nock.cleanAll();
+    if (clock) {
+      clock.restore();
+    }
+  });
+
+  describe('Constructor', () => {
+    it('should throw on empty access boundary rule', () => {
+      const expectedError = new Error(
+        'At least one access boundary rule needs to be defined.'
+      );
+      const cabWithEmptyAccessBoundaryRules = {
+        accessBoundary: {
+          accessBoundaryRules: [],
+        },
+      };
+      assert.throws(() => {
+        return new DownscopedClient(client, cabWithEmptyAccessBoundaryRules);
+      }, expectedError);
+    });
+
+    it('should throw on exceed number of access boundary rules', () => {
+      const expectedError = new Error(
+        'Access boundary rule exceeds limit, max 10 allowed.'
+      );
+      const cabWithExceedingAccessBoundaryRules = {
+        accessBoundary: {
+          accessBoundaryRules: [] as any,
```

Comment:
You don't need the `as any`

Reply:
I have to keep this, [] by default is type never,
Otherwise I have to export CAB interface and import it in the test.
I don't think it is good to export more unnecessary interfaces.